storage/innobase/buf/buf0flu.cc

```diff
@@ -2087,6 +2091,77 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
   return (true);
 }
 
+/** Waits until a flush batch of the given lsn ends
+@param[in]	new_oldest	target oldest_modified_lsn to wait for */
+static void buf_flush_wait_flushed(lsn_t new_oldest) {
+  lsn_t instance_newest[MAX_BUFFER_POOLS];
+  for (ulint i = 0; i < srv_buf_pool_instances; ++i) {
+    buf_pool_t *buf_pool;
+    instance_newest[i] = 0;
+
+    buf_pool = buf_pool_from_array(i);
+
+    if (new_oldest == LSN_MAX) {
+      buf_flush_list_mutex_enter(buf_pool);
+
+      buf_page_t *bpage = UT_LIST_GET_FIRST(buf_pool->flush_list);
+
+      while (bpage != NULL && fsp_is_system_temporary(bpage->id.space())) {
+        bpage = UT_LIST_GET_NEXT(list, bpage);
+      }
+
+      if (bpage != NULL) {
+        ut_ad(bpage->in_flush_list);
+        instance_newest[i] = bpage->oldest_modification;
+      }
```

Comment:
seems that oldest_modification is always the last from the list - shouldn't you go backward then and get first page that is not fsp_is_system_temporary ?